src/main/frontend/handler/common/file.cljs

```diff
@@ -19,20 +20,41 @@
       (when (not= file current-file)
         current-file))))
 
-(defn- get-delete-blocks [repo-url first-page file]
-  (let [delete-blocks (->
-                       (concat
-                        (db/delete-file-blocks! repo-url file)
-                        (when first-page (db/delete-page-blocks repo-url (:block/name first-page))))
-                       (distinct))]
+(defn- get-clear-blocks-tx
+  [blocks retain-uuids]
+  (let [tx-for-block (fn [block] (let [{uuid :block/uuid eid :db/id} block
+                                       should-retain? (and uuid (contains? retain-uuids uuid))]
+                                   (cond
+                                     should-retain?
+                                     (map (fn [attr] [:db.fn/retractAttribute eid attr]) db-schema/retract-attributes)
+                                     :else
+                                     [[:db.fn/retractEntity eid]])))]
+    (mapcat tx-for-block (distinct blocks)))
+  )
+
+(defn- get-clear-block-tx
+  "Returns the transactional operations to clear blocks belonging to the given
+  page and file.
+
+  Blocks are by default fully deleted via retractEntity. However, a collection
+  of block UUIDs to retain can be passed, and any blocks with matching uuids
+  will instead have their attributes cleared individually via
+  'retractAttribute'. This will preserve block references to the retained
+  UUIDs."
+  [repo-url first-page file retain-uuid-blocks]
+  (let [pages-to-clear (filter some? [(db/get-file-page file) file])
+        blocks (mapcat (fn [page] (db/get-page-blocks-no-cache repo-url page {:pull-keys [:db/id :block/uuid]})) pages-to-clear)
```

Comment:
@tiensonqin This seems to incorrectly get page blocks for a file and a page whereas the previous implementation got file blocks and page blocks. Fetching blocks both ways seems repetitive. In the previous implementation, is there a case where `db/delete-page-blocks` is getting a unique block that `db/delete-file-blocks!` doesn't? I saw that fn was introduced in https://github.com/logseq/logseq/commit/77ce2d8eaad06d77e2c6efd4d8078703730bb917 but didn't understand the bug it fixed

Reply:
@mrtracy I see file is passed to `db/get-page-blocks-no-cache` which always results in nil since it is not a page. It looks like you forgot to convert the equivalent of `db/delete-file-blocks!` to this new approach

Reply:
This was a mistake, but also I agree its a confusing situation.  I've corrected it so that it matches the original intention.

As I understand it, the page name can diverge from the file name via the 'title::' property.  Thus, the page name which comes from the newly parsed file might differ from the one on disk - therefore, the old logic was deleting blocks belonging to both pages.

However, I don't understand the circumstance where the new page name would both:
1. Already contain child blocks
2. Not be associated with another file already.

This method publishes a visible error if the new page name is already owned by another file, but doesn't appear to prevent it from happening - my change has not modified this.

In terms of using 'delete-page-blocks' vs 'delete-file-blocks!', well, 'delete-file-blocks!' just looks up the associated page block.  I thought that looking up the old page name made the intended purpose more apparent.

Note: I've added an additional test to exercise the "reload page with title field" case.